### DISCUSSION TITLE
Deleting joint in existing model and recreating?

### DISCUSSION THREAD

**Discussion by mhortho44** [ComputationalBiomechanicsLab/opensim-creator#930] [Q&A] [answered]
How would one go about deleting or modifying a joint in an existing model?

Specifically for the purposes of changing the connections of the joint in order to recreate it? When I attempt to do this I get different error messages.

i.e. Thoracoscapular model (Seth) - Delete GH joint and then recreate joints in OSC UI in order to construct a reverse shoulder arthroplasty:
RSA components needed to be represented:
(Glenoid-baseplate-glenosphere (ball joint) humerosocket-humeral bone) "-" = weldjoint/no motion interface

I think the topology would be something like this:

Scapula(body) -> Scapula_offset -> scapula_to_baseplate (weldjoint)-> baseplate_offset -> baseplate(Body) -> baseplate_offset #Representing the actual lateralization -> baseplate_to_glenosphere (Weldjoint) -> glenosphere_offset-> glenosphere (body)-> glenosphere_offset-> glenohumeral_joint (Ball joint) -> humeral_component_offset -> Humeral_component_contactpoint (body)-> Hum_component_offset (#Variable of interest humerus offset/distalization)->humerus_offset-> humerus (body)

The reason for all of the weldjoint redundancies would be to make it easier to measure clinically relevant offset lengths (lateralization and distalization) as they are conventionally measured intra-operatively.

I can even create this series of connections in model but without being able to remove the existing "complex" Glenohumeral joint, it doesn't behave properly.

**Comment by adamkewley** [comment, marked as answer]
## Joint Deletion

Deleting a joint is currently not supported in OSC. It's something people want, and we're aware it's an issue. It's held up by the fact that deletion logic has to (e.g.) resolve kinematic chains, sockets, model hierarchies, etc. It would also need to ask the user, probably via a popup, how to handle any ambiguities that were automatically detected by such an algorithm (e.g. by asking the user whether they want to reconnect a child of the joint to something else or delete the entire child chain, and so on).

The "easiest" (not easy) way to delete a joint from a model is to manually slice it out of the XML file. It's do-able, but takes a bit of work because you'll inevitably find that child bodies, joints, etc. need to be handled correctly before OSC/OpenSim are willing to open the osim file. There's an `Export Model Graph as Dotviz (clipboard)` option under `Tools > Experimental Tools` in the editor UI, you can paste the dotviz markup [here](https://dreampuf.github.io/GraphvizOnline). This feature tends to give users anxiety.

The reason you cannot remove the "complex" glenohumeral joint without "something behaving properly" is probably because there are components in the model that depend on the existence of the joint. This isn't limited to child bodies/other joints. E.g. any muscle paths that contain points that are defined in bodies that are children of the glenohumeral joint are indirectly dependent on the existence of the joint. Hopefully, you can see why handling this automatically inside the UI while guaranteeing no crashes, undo/redo support, etc. is a little complicated :wink: .

## Joint Modification

For modifying/changing the connections of an existing joint, your best bet in OSC is currently the `Sockets` option that appears when you right-click an `OpenSim::Joint` in the model editor. This lets you reassign the `parent_frame` or `child_frame` sockets, which effectively lets you attach the joint to something else.

If the frame being reassigned is a `PhysicalOffsetFrame` then a `Re-express $FRAME in chosen frame` checkbox is also added to the UI, which should perform the necessary maths to ensure that the reattachment doesn't move the joint in the global reference frame (otherwise, it will snap to whatever it is reassigned to).

There's also a `Change Joint Type` option in the right-click menu. This is sometimes handy for various little model hacks (e.g. using IK to fit to a looser version of the joint before constraining it).

## Weld Joint Usage

I'm unsure about your logic w.r.t. weld joints. They don't have any outputs that would give you anything more clinically relevant than (e.g.) measuring the distance between the parent/child frames. I'm unsure, but you can probably simplify what you're doing to:

- Scapula(body) -> Scapula_offset -> baseplate_offset -> baseplate(Body) -> baseplate_offset

You'd still end up with two coordinate systems (`Scapula_offset` and `baseplate_offset`), but without also having a joint hanging around. However, having the joint can be handy later on if you (e.g.) decide you want to swap it for a `FreeJoint` or `PinJoint` or similar - if you think that'd be handy for simulation.

**Comment by itbellix** [comment]
Hi @adamkewley, I know I am reviving an old discussion here, so hopefully that's alright. I was just trying to use the `Change Joint Type` functionality on my model to transform a `CustomJoint` into a `WeldJoint`, and it works well for me (with version 0.5.17, and after having deleted the coordinate actuators that were related to the previous mobile joint). What I found however is that this process welds the joint in its original zero-coordinate position, and I am not able to lock the joint at a custom position. Is this something that would make sense as a feature request?

Thank you again for all the great work!

**Reply by adamkewley** [reply]
@itbellix changing the joint type will use the zero-coordinate value by default because the operation literally just changes the joint type and doesn't attempt to figure out how one coordinate's constraint manifold may project onto another type of joint. In the case of a `WeldJoint` there is, of course, only one coordinate value (nothing).

However, for your particular requirements, you might be able to use the `Rezero Joint` menu option that's directly below the `Change Joint Type` menu option in the editor? The operation requires that the joint is attached via two `PhysicalOffsetFrame`s either side of it, and works by changing those offset frames' transformations such that the current joint's zero point produces the same transformation as the pre-zeroed coordinate value. Once zeroed, the joint type can be changed.

**Reply by itbellix** [reply]
@adamkewley thanks for the explanation! I will give it a try :)
